src/librustc_codegen_llvm/type_of.rs

```diff
@@ -52,32 +52,35 @@ fn uncached_llvm_type<'a, 'tcx>(
     }
 
     let name = match layout.ty.kind {
-        ty::Closure(..) |
-        ty::Generator(..) |
-        ty::Adt(..) |
-        // FIXME(eddyb) producing readable type names for trait objects can result
+        ty::Closure(..) | ty::Generator(..) | ty::Adt(..) | ty::Foreign(..) | ty::Str
```

Comment:
Not sure duplicating this list is worth it. Also, none of closures, generators, foreign types or `str` can be recursive AFAIK (as in, none of their fields can end up referring back to them), so we could just use `None` for them.

So effectively, I'm suggesting adding `if !cx.sess().fewer_names()` to the arm that actually generates a name, and then having `ty::Adt(..) => Some(String::new())` after it.

Reply:
I'll make this change after the perf run is in.

Reply:
Fixed!